File: tests/layer0_ops/unit/TestFFTFull.m
classdef TestFFTFull < matlab.unittest.TestCase
    % TestFFTFull: unit tests for comp_filterbank_fft and comp_ifilterbank_fft
    %
    % Correct calling conventions (from comp_filterbank.m):
    %   c = comp_filterbank_fft(F, G, a)
    %       F : fft(f), size [L x W]          -- FFT of the signal, NOT the signal itself
    %       G : cell{M} of length-L DFT responses
    %       a : integer subsampling vector, a(m) must divide L exactly
    %
    %   F = comp_ifilterbank_fft(c, G, a)
    %       c : cell{M} of coefficient matrices [N(m) x W]
    %       G : same cell of length-L DFT responses
    %       a : same subsampling vector
    %       (no Ls argument -- output length = numel(G{1}))

    properties
        Ls           % signal length
        M            % number of subbands
        a            % subsampling factors  [M x 1]
        G            % cell{M} of rectangular bandpass DFT responses (length Ls)
        noise_real
        noise_stereo
        zeros_sig
    end

    methods(TestClassSetup)
        function setupTestClass(testCase)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), ...
                                 '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
            testCase.Ls = 1024;
            testCase.M  = 4;
            testCase.a  = [4; 8; 8; 16];   % subsampling factors (all divide 1024)

            % Build M rectangular full-length DFT responses
            Ls    = testCase.Ls;
            M     = testCase.M;
            bw    = floor(Ls / (2 * M + 2));   % bandwidth per band (in bins)
            G_tmp = cell(M, 1);
            for m = 1 : M
                g_full  = zeros(Ls, 1);
                fc_bin  = round(m * Ls / (2 * (M + 1)));
                lo = max(1, fc_bin - floor(bw/2));
                hi = min(Ls, fc_bin + floor(bw/2));
                g_full(lo : hi) = 1;
                G_tmp{m} = g_full;
            end
            testCase.G = G_tmp;

            testCase.noise_real   = randn(testCase.Ls, 1);
            testCase.noise_stereo = randn(testCase.Ls, 2);
            testCase.zeros_sig    = zeros(testCase.Ls, 1);
        end
    end

    methods(Test)

        function testFFTFilterbankZeroInput(testCase)
            % Zero signal -> all subbands zero
            F = fft(testCase.zeros_sig);
            c = comp_filterbank_fft(F, testCase.G, testCase.a);

            for m = 1:testCase.M
                testCase.verifyEqual(c{m}, zeros(size(c{m})), 'AbsTol', 1e-14, ...
                    sprintf('Subband %d not zero for zero input', m));
            end
        end

        function testFFTFilterbankOutputSizes(testCase)
            % Each c{m} has N(m) = Ls/a(m) rows
            F = fft(testCase.noise_real);
            c = comp_filterbank_fft(F, testCase.G, testCase.a);

            for m = 1:testCase.M
                expected = testCase.Ls / testCase.a(m);
                testCase.verifyEqual(size(c{m}, 1), expected, ...
                    sprintf('Subband %d: expected %d rows, got %d', m, expected, size(c{m},1)));
            end
        end

        function testFFTFilterbankRealNoise(testCase)
            % Real noise: M subbands returned
            F = fft(testCase.noise_real);
            c = comp_filterbank_fft(F, testCase.G, testCase.a);
            testCase.verifyEqual(length(c), testCase.M);
        end

        function testFFTFilterbankMultiChannel(testCase)
            % Multi-channel [Ls x 2]: each c{m} has 2 columns
            F = fft(testCase.noise_stereo);
            c = comp_filterbank_fft(F, testCase.G, testCase.a);

            for m = 1:testCase.M
                testCase.verifyEqual(size(c{m}, 2), 2, ...
                    sprintf('Subband %d: expected 2 columns, got %d', m, size(c{m},2)));
            end
        end

        function testFFTFilterbankIntegerSubsampling(testCase)
            % All a(m) divide Ls exactly (required by comp_filterbank_fft)
            testCase.verifyTrue(all(mod(testCase.Ls, testCase.a) == 0));
        end

        function testInverseFFTFilterbankZeroInput(testCase)
            % Zero coefficients -> zero output from synthesis.
            % comp_ifilterbank_fft requires numel(c{m}) = Ls/a(m) so that
            % repmat(fft(c{m}), a(m), 1) tiles to exactly Ls samples.
            % Each channel therefore needs its own length Ls/a(m).
            c_zero = cell(testCase.M, 1);
            for m = 1:testCase.M
                N_m = testCase.Ls / testCase.a(m);
                c_zero{m} = zeros(N_m, 1);
            end
            F_recon = comp_ifilterbank_fft(c_zero, testCase.G, testCase.a);
            testCase.verifyEqual(F_recon, zeros(testCase.Ls, 1), 'AbsTol', 1e-10);
        end

        function testForwardInverseRoundtrip(testCase)
            % For the trivial M=1 all-pass filter (G{1} = ones(L,1), a=1):
            %   comp_filterbank_fft(fft(f), G, a)    -> c{1} = f
            %   comp_ifilterbank_fft(c, G, a)        -> F_recon = fft(f)
            % so ifft(F_recon) = f exactly.
            G_trivial = {ones(testCase.Ls, 1)};
            a_trivial = ones(1, 1);   % subsampling = 1
            F       = fft(testCase.noise_real);
            c       = comp_filterbank_fft(F, G_trivial, a_trivial);
            F_recon = comp_ifilterbank_fft(c, G_trivial, a_trivial);
            f_recon = real(ifft(F_recon));

            err = norm(f_recon - testCase.noise_real) / norm(testCase.noise_real);
            testCase.verifyLessThan(err, 1e-10, ...
                'All-pass roundtrip: ifft(F_recon) should equal f');
        end

    end

end

File: tests/layer0_ops/unit/test_fft_full.py
"""
test_fft_full.py
================
Python port of:
    layer0_ops/unit/TestFFTFull.m

Covers: comp_filterbank_fft, comp_ifilterbank_fft

Calling conventions
-------------------
comp_filterbank_fft(F, G, a)
    F : np.ndarray [L] or [L x W] — fft(f), NOT the signal itself
    G : list of M length-L DFT responses
    a : array-like of M integer subsampling factors (all must divide L)

comp_ifilterbank_fft(c, G, a)
    c : list of M coefficient arrays [N_m] or [N_m x W], N_m = L / a_m
    G : same list of length-L DFT responses
    a : same subsampling factors
    Returns: length-L frequency-domain array (apply ifft to get time domain)

The fixture fft_fb (from conftest) provides:
    Ls, M, a, G, noise_real, noise_stereo, zeros_sig
"""

from __future__ import annotations

import pytest

import numpy as np

# ---------------------------------------------------------------------------
# Reference tests (no impl required) – structural / algebraic checks
# using the synthetic fixture built in conftest
# ---------------------------------------------------------------------------

class TestFFTFilterbankReference:
    """
    Structural checks that can be done with pure numpy using the fixture data.
    MATLAB counterpart: TestFFTFull.
    """

    def test_subsampling_divides_Ls(self, fft_fb):
        """All a(m) must divide Ls exactly (required by comp_filterbank_fft)."""
        Ls = fft_fb["Ls"]
        for a_m in fft_fb["a"]:
            assert Ls % a_m == 0, f"a={a_m} does not divide Ls={Ls}"

    def test_filter_lengths(self, fft_fb):
        """Each G[m] has length Ls."""
        Ls = fft_fb["Ls"]
        for m, g in enumerate(fft_fb["G"]):
            assert len(g) == Ls, f"G[{m}] has length {len(g)}, expected {Ls}"

    def test_num_bands(self, fft_fb):
        """Number of filters equals M."""
        assert len(fft_fb["G"]) == fft_fb["M"]

    def test_rectangular_filters_are_binary(self, fft_fb):
        """Synthetic rectangular filters contain only 0 and 1."""
        for g in fft_fb["G"]:
            assert set(np.unique(g)).issubset({0.0, 1.0})

    def test_noise_signal_length(self, fft_fb):
        """noise_real and zeros_sig have length Ls."""
        assert len(fft_fb["noise_real"]) == fft_fb["Ls"]
        assert len(fft_fb["zeros_sig"]) == fft_fb["Ls"]


# ---------------------------------------------------------------------------
# comp_filterbank_fft – implementation tests
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestCompFilterbankFFTImpl:
    """
    MATLAB counterpart: TestFFTFull (comp_filterbank_fft section).
    """

    def test_zero_input(self, needs_impl, fft_fb):
        """Zero signal -> all subbands zero."""
        from cool_frames.core import comp_filterbank_fft  # type: ignore

        F = np.fft.fft(fft_fb["zeros_sig"])
        c = comp_filterbank_fft(F, fft_fb["G"], fft_fb["a"])

        for m, c_m in enumerate(c):
            np.testing.assert_allclose(
                c_m, np.zeros_like(c_m), atol=1e-14,
                err_msg=f"Subband {m} not zero for zero input"
            )

    def test_output_sizes(self, needs_impl, fft_fb):
        """Each c[m] has N_m = Ls/a[m] rows."""
        from cool_frames.core import comp_filterbank_fft  # type: ignore

        Ls = fft_fb["Ls"]
        F  = np.fft.fft(fft_fb["noise_real"])
        c  = comp_filterbank_fft(F, fft_fb["G"], fft_fb["a"])

        for m, (c_m, a_m) in enumerate(zip(c, fft_fb["a"])):
            expected = Ls // a_m
            assert c_m.shape[0] == expected, \
                f"Subband {m}: expected {expected} rows, got {c_m.shape[0]}"

    def test_real_noise_returns_M_bands(self, needs_impl, fft_fb):
        """Real noise: M subbands returned."""
        from cool_frames.core import comp_filterbank_fft  # type: ignore

        F = np.fft.fft(fft_fb["noise_real"])
        c = comp_filterbank_fft(F, fft_fb["G"], fft_fb["a"])
        assert len(c) == fft_fb["M"]

    def test_multichannel(self, needs_impl, fft_fb):
        """Multi-channel [Ls x 2]: each c[m] has 2 columns."""
        from cool_frames.core import comp_filterbank_fft  # type: ignore

        F = np.fft.fft(fft_fb["noise_stereo"], axis=0)
        c = comp_filterbank_fft(F, fft_fb["G"], fft_fb["a"])

        for m, c_m in enumerate(c):
            assert c_m.shape[1] == 2, \
                f"Subband {m}: expected 2 columns, got {c_m.shape[1]}"


# ---------------------------------------------------------------------------
# comp_ifilterbank_fft – implementation tests
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestCompIFilterbankFFTImpl:
    """
    MATLAB counterpart: TestFFTFull (comp_ifilterbank_fft section).
    """

    def test_zero_coefficients(self, needs_impl, fft_fb):
        """Zero coefficients -> zero frequency-domain output."""
        from cool_frames.core import comp_ifilterbank_fft  # type: ignore

        Ls     = fft_fb["Ls"]
        c_zero = [np.zeros(Ls // a_m) for a_m in fft_fb["a"]]
        F_recon = comp_ifilterbank_fft(c_zero, fft_fb["G"], fft_fb["a"])

        np.testing.assert_allclose(F_recon, np.zeros(Ls), atol=1e-10)

    def test_allpass_roundtrip(self, needs_impl, fft_fb):
        """All-pass filter (G=[1,...,1], a=1): ifft(ifilterbank(filterbank(F))) ≈ f."""
        from cool_frames.core import comp_filterbank_fft, comp_ifilterbank_fft  # type: ignore

        Ls        = fft_fb["Ls"]
        f         = fft_fb["noise_real"]
        G_trivial = [np.ones(Ls, dtype=complex)]
        a_trivial = np.array([1])

        F       = np.fft.fft(f)
        c       = comp_filterbank_fft(F, G_trivial, a_trivial)
        F_recon = comp_ifilterbank_fft(c, G_trivial, a_trivial)
        f_recon = np.real(np.fft.ifft(F_recon))

        err = np.linalg.norm(f_recon - f) / np.linalg.norm(f)
        assert err < 1e-10, \
            f"All-pass roundtrip: relative error {err:.2e} exceeds 1e-10"


# ---------------------------------------------------------------------------
# combined forward+inverse – implementation test
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFFTRoundtripImpl:
    """Forward + inverse roundtrip consistency checks."""

    def test_forward_inverse_consistency(self, needs_impl, fft_fb):
        """comp_ifilterbank_fft(comp_filterbank_fft(F, G, a), G, a) is non-trivial."""
        from cool_frames.core import comp_filterbank_fft, comp_ifilterbank_fft  # type: ignore

        F       = np.fft.fft(fft_fb["noise_real"])
        c       = comp_filterbank_fft(F, fft_fb["G"], fft_fb["a"])
        F_recon = comp_ifilterbank_fft(c, fft_fb["G"], fft_fb["a"])

        # Output should have the right length and non-trivial energy
        assert len(F_recon) == fft_fb["Ls"]
        assert np.linalg.norm(F_recon) > 0
